src/NHibernate/Driver/MicrosoftDataSqlClientDriver.cs

```diff
@@ -0,0 +1,206 @@
+
+		private Dialect.Dialect _dialect;
+
+		public MicrosoftDataSqlClientDriver()
+			: base(
+				"Microsoft.Data.SqlClient",
+				"Microsoft.Data.SqlClient.SqlConnection",
+				"Microsoft.Data.SqlClient.SqlCommand")
+		{
+		}
+
+		/// <summary>
+		///     MsSql requires the use of a Named Prefix in the SQL statement.
+		/// </summary>
+		/// <remarks>
+		///     <see langword="true" /> because MsSql uses "<c>@</c>".
+		/// </remarks>
+		public override bool UseNamedPrefixInSql => true;
+
+		/// <summary>
+		///     MsSql requires the use of a Named Prefix in the Parameter.
+		/// </summary>
+		/// <remarks>
+		///     <see langword="true" /> because MsSql uses "<c>@</c>".
+		/// </remarks>
+		public override bool UseNamedPrefixInParameter => true;
+
+		/// <summary>
+		///     The Named Prefix for parameters.
+		/// </summary>
+		/// <value>
+		///     Sql Server uses <c>"@"</c>.
+		/// </value>
+		public override string NamedPrefix => "@";
+
+		/// <inheritdoc/>
+		public override bool SupportsMultipleOpenReaders => true;
+
+		/// <inheritdoc />
+		public override bool SupportsMultipleQueries => true;
+
+		/// <summary>
+		///     With read committed snapshot or lower, SQL Server may have not actually already committed the transaction
+		///     right after the scope disposal.
+		/// </summary>
+		public override bool HasDelayedDistributedTransactionCompletion => true;
+
+		public override bool RequiresTimeSpanForTime => true;
+
+		/// <inheritdoc />
+		public override DateTime MinDate => DateTime.MinValue;
+
+		System.Type IEmbeddedBatcherFactoryProvider.BatcherFactoryClass => typeof(GenericBatchingBatcherFactory);
+
+		/// <inheritdoc />
+		public virtual void AdjustParameterForValue(DbParameter parameter, SqlType sqlType, object value)
+		{
+			if (value is string stringVal)
+				switch (parameter.DbType)
+				{
+					case DbType.AnsiString:
+					case DbType.AnsiStringFixedLength:
+						parameter.Size = IsAnsiText(parameter, sqlType)
+							? MsSql2000Dialect.MaxSizeForAnsiClob
+							: Math.Max(stringVal.Length, sqlType.LengthDefined ? sqlType.Length : parameter.Size);
+						break;
+					case DbType.String:
+					case DbType.StringFixedLength:
+						parameter.Size = IsText(parameter, sqlType)
+							? MsSql2000Dialect.MaxSizeForClob
+							: Math.Max(stringVal.Length, sqlType.LengthDefined ? sqlType.Length : parameter.Size);
+						break;
+				}
+		}
+
+		public override void Configure(IDictionary<string, string> settings)
+		{
+			base.Configure(settings);
+
+			_dialect = Dialect.Dialect.GetDialect(settings);
+		}
+
+		/// <inheritdoc />
+		protected override void InitializeParameter(DbParameter dbParam, string name, SqlType sqlType)
+		{
+			base.InitializeParameter(dbParam, name, sqlType);
+
+			// Defaults size/precision/scale
+			switch (dbParam.DbType)
```

Comment:
There is a subtle difference in our Sql2008 driver here : it reads `sqlType.DbType`, not the `dbParam` one. And due to an [old legacy](https://github.com/dotnet/SqlClient/issues/5) they have unfortunately kept in Microsoft.Data.SqlClient, this changes a lot: `DbType.Date` and `DbType.Time` are changed to `DbType.DateTime` on the parameter.
So it is critical to read `sqlType.DbType` here instead.

That is the reason being the test failures I was reporting.

Reply:
Fixed